File: python/widgets/keypad.py
from PySide6.QtWidgets import QWidget, QGridLayout, QPushButton, QVBoxLayout, QLineEdit
from PySide6.QtCore import Qt, Signal
import theme

class TacticalKeypad(QWidget):
    """
    A touch-optimized numerical keypad designed to fit inside 
    a TacticalPopout content area.
    """
    # Signal emitted when 'ENTER' is pressed
    submitted = Signal(str)
    
    def __init__(self, initial_text="", parent=None):
        super().__init__(parent)
        self.layout = QVBoxLayout(self)
        self.layout.setContentsMargins(0, 0, 0, 0)
        self.layout.setSpacing(10)

        # 1. Preview Display (Shows what the user is typing)
        self.display = QLineEdit(str(initial_text))
        self.display.setAlignment(Qt.AlignCenter)
        self.display.setReadOnly(True)
        self.refresh_display_style()
        self.layout.addWidget(self.display)

        # 2. Key Grid
        self.grid = QGridLayout()
        self.grid.setSpacing(6)
        
        # Numpad Layout optimized for Deadzone percentages (0-100)
        keys = [
            ['7', '8', '9'],
            ['4', '5', '6'],
            ['1', '2', '3'],
            ['CLR', '0', 'BACK'],
            ['ENTER']
        ]

        for r, row in enumerate(keys):
            for c, key in enumerate(row):
                btn = QPushButton(key)
                
                # 'ENTER' button spans the entire bottom row
                col_span = 3 if key == 'ENTER' else 1
                
                btn.setFixedHeight(50)
                btn.setStyleSheet(self._get_key_style(key))
                
                # FIX: Added '_' to catch the boolean 'checked' signal from the click event
                # This prevents the TypeError: can only concatenate str (not "bool") to str
                btn.clicked.connect(lambda _, k=key: self._on_key_press(k))
                
                self.grid.addWidget(btn, r, c, 1, col_span)

        self.layout.addLayout(self.grid)

    def _get_key_style(self, key):
        h = theme.ACTIVE['hex']
        bg = "rgba(45, 45, 45, 0.9)"
        txt = "white"
        
        if key == 'ENTER':
            bg = h
            txt = "black"
        
        return f"""
            QPushButton {{
                background: {bg};
                color: {txt};
                border: 1px solid rgba(255,255,255,0.1);
                font-family: 'Consolas';
                font-weight: bold;
                font-size: 16px;
                border-radius: 4px;
            }}
            QPushButton:pressed {{
                background: {h};
                color: black;
            }}
        """

    def _on_key_press(self, key):
        curr = self.display.text()
        
        if key == 'ENTER':
            # Send result back to the popout
            self.submitted.emit(curr)
            
        elif key == 'BACK':
            # Remove last character
            self.display.setText(curr[:-1])
            
        elif key == 'CLR':
            # Clear everything
            self.display.clear()
            
        else:
            # Numerical input: Limit to 3 digits (max 100 for deadzones)
            if len(curr) < 3:
                # Ensure we only handle string addition
                self.display.setText(curr + str(key))

    def refresh_display_style(self):
        """Helper to set/update the display styling based on active theme."""
        h = theme.ACTIVE['hex']
        self.display.setStyleSheet(f"""
            QLineEdit {{
                background: rgba(0, 0, 0, 0.6);
                border: 1px solid {h};
                color: white;
                font-family: 'Consolas';
                font-size: 24px;
                padding: 8px;
                border-radius: 4px;
            }}
        """)
                
    def refresh_theme(self):
        """Updates the keypad visuals live if the theme changes."""
        self.refresh_display_style()
        # Refresh all buttons in the grid to update colors
        for i in range(self.grid.count()):
            widget = self.grid.itemAt(i).widget()
            if isinstance(widget, QPushButton):
                widget.setStyleSheet(self._get_key_style(widget.text()))

File: src/GUI/widgets/keypad.cpp
#include "theme.h"
#include "widgets/keypad.h"
#include <QVBoxLayout>
#include <QGridLayout>
#include <QLineEdit>
#include <QTimer>

Keypad::Keypad(double initial_val, QWidget *parent) : QWidget(parent) {
    setAttribute(Qt::WA_AcceptTouchEvents);
    setFocusPolicy(Qt::StrongFocus);

    QVBoxLayout *layout = new QVBoxLayout(this);
    layout->setContentsMargins(10, 10, 10, 10);
    layout->setSpacing(8);

    display = new QLineEdit();
    display->setAlignment(Qt::AlignCenter);
    display->setReadOnly(true);
    display->setFixedHeight(50);
    layout->addWidget(display);

    QGridLayout *grid = new QGridLayout();
    grid->setSpacing(8);

    QStringList keys_text = {"7", "8", "9", "4", "5", "6", "1", "2", "3", ".", "0", "-", "CLR", "BACK", "CANCEL", "ENTER"};
    
    int row = 0, col = 0;
    for (const QString &txt : keys_text) {
        Button *btn = new Button(txt, this, true);
        btn->setFixedSize(70, 60);
        if (txt == "ENTER") {
            grid->addWidget(btn, row, col, 1, 2);
            btn->setFixedSize(148, 60);
        } else {
            grid->addWidget(btn, row, col);
        }
        connect(btn, &QPushButton::clicked, this, &Keypad::handle_key);
        keys.append(btn);
        col++;
        if (col > 2 && txt != "ENTER") { col = 0; row++; }
        if (txt == "ENTER") { col = 0; row++; }
    }
    layout->addLayout(grid);
    refresh_theme();
}

void Keypad::reset() {
    display->clear(); // Ensure it is blank on every open
    display->setFocus();
}

void Keypad::handle_key() {
    Button *btn = qobject_cast<Button *>(sender());
    if (!btn) return;
    QString t = btn->text();
    QString cur = display->text();

    if (t == "ENTER") emit submitted(cur.toDouble());
    else if (t == "CANCEL") emit cancelled();
    else if (t == "CLR") display->clear();
    else if (t == "BACK") display->setText(cur.left(cur.length() - 1));
    else if (t == "-") {
        if (cur.startsWith("-")) display->setText(cur.mid(1));
        else display->setText("-" + cur);
    }
    else display->insert(t);
}

void Keypad::refresh_theme() {
    QString h = ThemeManager::instance().active().hex;
    display->setStyleSheet(QString("background:#000; color:white; border:1px solid %1; font-size:24px;").arg(h));
    for (Button *b : keys) {
        if (b->text() == "ENTER") b->setStyleSheet(QString("QPushButton{background:%1; color:black; font-weight:bold;}").arg(h));
        if (b->text() == "CANCEL") b->setStyleSheet("QPushButton{background:#441111; color:white;}");
    }
}
